common/src/main/java/com/microsoft/identity/common/adal/internal/cache/StorageHelper.java

```diff
@@ -220,42 +231,77 @@ public String encrypt(final String clearText)
     }
 
     @Override
-    public String decrypt(final String encryptedBlob)
-            throws GeneralSecurityException, IOException {
+    public String decrypt(final String encryptedBlob) throws GeneralSecurityException, IOException {
+        final String methodName = "decrypt";
         Log.v(TAG, "Starting decryption");
 
         if (StringExtensions.isNullOrBlank(encryptedBlob)) {
             throw new IllegalArgumentException("Input is empty or null");
         }
 
         int encodeVersionLength = encryptedBlob.charAt(0) - 'a';
-        if (encodeVersionLength <= 0) {
-            throw new IllegalArgumentException(String.format(
-                    "Encode version length: '%s' is not valid, it must be greater of equal to 0",
-                    encodeVersionLength));
-        }
-        if (!encryptedBlob.substring(1, 1 + encodeVersionLength).equals(ENCODE_VERSION)) {
-            throw new IllegalArgumentException(String.format(
-                    "Encode version received was: '%s', Encode version supported is: '%s'", encryptedBlob,
-                    ENCODE_VERSION));
-        }
+        validateEncodeVersion(encryptedBlob, encodeVersionLength);
+
+        final byte[] bytes = Base64.decode(
+                encryptedBlob.substring(1 + encodeVersionLength),
+                Base64.DEFAULT
+        );
 
-        final byte[] bytes = Base64
-                .decode(encryptedBlob.substring(1 + encodeVersionLength), Base64.DEFAULT);
+        final String packageName = mContext.getPackageName();
 
+        try {
+            return decryptWithPackageName(bytes, packageName);
+        } catch (GeneralSecurityException | IOException e) {
+            Logger.error(
+                    TAG + methodName,
+                    "Failed to decrypt with package name: "
+                            + packageName,
+                    e
+            );
+
+            if (!BROKER_PACKAGE_NAMES.contains(packageName)) {
+                // If we are not the broker, do not retry, immediately throw.
+                throw e;
```

Comment:
This check can be moved to right after line 250 (final String packageName = mContext.getPackageName();) - we should not even attempting to read if we are not one of these 2 apps.

Reply:
Not true; this is shared code with MSAL